File: src/linked/copy_list_with_random.java
package linked;

import java.util.HashMap;

class NodeR {
  int val;
  NodeR next;
  NodeR random;

  public NodeR(int val) {
    this.val = val;
    this.next = null;
    this.random = null;
  }
}

class Solution6 {
  public NodeR copyRandomList1(NodeR head) {
    HashMap<NodeR, NodeR> map = new HashMap<>();// key 老节点；value 新节点
    NodeR cur = head;
    while (cur != null) {
      map.put(cur, new NodeR(cur.val));
      cur = cur.next;
    }
    cur = head;
    while (cur != null) {
      /**
       * cur 老
       * map.get(cur) 新
       * 新.next -> cur.next克隆节点找到
       */
      map.get(cur).next = map.get(cur.next);
      map.get(cur).random = map.get(cur.random);
      cur = cur.next;
    }
    return map.get(head);
  }

  public NodeR copyRandomList2(NodeR head) {
    if (head == null) {
      return null;
    }
    NodeR cur = head;
    NodeR next = null;
    while (cur != null) {
      next = cur.next;
      cur.next = new NodeR(cur.val);
      cur.next.next = next;
      cur = next;
    }
    cur = head;
    NodeR copy = null;
    // 依次设置 1' 2' 3' random指针
    while (cur != null) {
      next = cur.next.next;
      copy = cur.next;
      copy.random = cur.random != null ? cur.random.next : null;
      cur = next;
    }
    NodeR res = head.next;
    cur = head;

    // 分离新老链表
    while (cur != null) {
      next = cur.next.next;
      copy = cur.next;
      cur.next = next;
      copy.next = next != null ? next.next : null;
      cur = next;
    }
    return res;
  }

}

File: src/linked/copy_list_with_random.cpp
#include <iostream>
#include <unordered_map>

using namespace std;
struct Node {
  int value;
  Node *next, *random;
  Node(int value) : value(value), next(nullptr), random(nullptr){};
};
class Solution {
  Node* copyRandomList1(Node* head) {
    unordered_map<Node*, Node*> m;
    Node* cur = head;
    while (cur != nullptr) {
      m.insert(cur, new Node(cur->value));
      cur = cur->next;
    }

    cur = head;
    while (cur != nullptr) {
      m[cur]->next = m[cur->next];
      m[cur]->random = m[cur->random];
      cur = cur->next;
    }
    return m[head];
  }

  Node* copyRandomList2(Node* head) {
    if (head == nullptr) return nullptr;
    Node *cur = head, *next = nullptr;
    while (cur != nullptr) {
      next = cur->next;
      cur->next = new Node(cur->value);
      cur->next->next = next;
      cur = next;
    }

    cur = head;
    Node* copy = nullptr;
    // 依次设置 1' 2' 3' random指针
    while (cur != nullptr) {
      next = cur->next->next;
      copy = cur->next;
      copy->random = cur->random != nullptr ? cur->random->next : nullptr;
      cur = next;
    }
    Node* res = head->next;
    cur = head;

    // 分离新老链表
    while (cur != nullptr) {
      next = cur->next->next;
      copy = cur->next;
      cur->next = next;
      copy->next = next != nullptr ? next->next : nullptr;
      cur = next;
    }
    return res;
  }
};

int main() { return 0; }